SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracerUserInterface.cpp

```diff
@@ -0,0 +1,175 @@
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "Interop/Connection/SpatialEventTracer.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Interop/SpatialClassInfoManager.h"
+#include "SpatialView/EntityComponentId.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracerUserInterface);
+
+FString USpatialEventTracerUserInterface::CreateSpanId(UObject* WorldContextObject)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return {};
+	}
+
+	return SpatialGDK::SpatialEventTracer::SpanIdToString(EventTracer->CreateSpan().GetValue());
+}
+
+FString USpatialEventTracerUserInterface::CreateSpanIdWithCauses(UObject* WorldContextObject, const TArray<FString>& Causes)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return {};
+	}
+
+	TArray<Trace_SpanId> SpanIdS;
+	for (const FString& String : Causes)
+	{
+		SpanIdS.Add(SpatialGDK::SpatialEventTracer::StringToSpanId(String));
+	}
+
+	return SpatialGDK::SpatialEventTracer::SpanIdToString(EventTracer->CreateSpan(SpanIdS.GetData(), SpanIdS.Num()).GetValue());
+}
+
+void USpatialEventTracerUserInterface::TraceEvent(UObject* WorldContextObject, const FString& SpanId, FSpatialTraceEvent SpatialTraceEvent)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr)
+	{
+		return;
+	}
+
+	EventTracer->TraceEvent(SpatialTraceEvent, SpatialGDK::SpatialEventTracer::StringToSpanId(SpanId));
+}
+
+void USpatialEventTracerUserInterface::SetActiveSpanId(UObject* WorldContextObject, FEventTracerDynamicDelegate Delegate,
+													   const FString& SpanId)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return;
+	}
+
+	EventTracer->SpanIdStack.AddNewLayer(SpatialGDK::SpatialEventTracer::StringToSpanId(SpanId));
+	Delegate.Execute();
+	EventTracer->SpanIdStack.PopLayer();
+}
+
+void USpatialEventTracerUserInterface::AddSpanIdToStack(UObject* WorldContextObject, const FString& SpanId)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return;
+	}
+
+	EventTracer->SpanIdStack.AddNewLayer(SpatialGDK::SpatialEventTracer::StringToSpanId(SpanId));
+}
+
+void USpatialEventTracerUserInterface::PopSpanIdFromStack(UObject* WorldContextObject)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return;
+	}
+
+	EventTracer->SpanIdStack.PopLayer();
+}
+
+void USpatialEventTracerUserInterface::AddLatentActorSpanId(UObject* WorldContextObject, const AActor& Actor, const FString& SpanId)
```

Comment:
Maybe this could be TraceProperty and we just take the UProperty and work out the rest? In the future we may have more granular control over the relationship between properties/components (like being able to attach a span to a property directly)

Also do RepNotifies work with this?

Reply:
Maybe TraceModification might make more sense here?

Reply:
Hmm, so UProperty is depreciated for FProperty. FProperty can not be used directly in blueprints as it is not a uclass or ustruct. If it was, how would you connect the property node? Does that work?

Reply:
All in favor of a new function name :P

Reply:
Oh fair enough, I didn't realise that. I guess right now we only trace at the entity/component level anyway